### DISCUSSION TITLE
Calculation of a value on a boundary surface

### DISCUSSION THREAD

**Discussion by Joe199999** [fppimenta/rheoTool#130] [General]
Hi 
I want to calculate the value of the stress tensor in cylindrical coordinates on a boundary surface. I did it in 2 ways:

1. Calculation the stress tensor (tet3) values on the surface and transfer them to cylindrical coordinates

`List<symmTensor> tet30 = tet3.boundaryField()[patchI];
List<symmTensor> tet3RT1 = Tau0;  // Initialization
tet3RT1[i].xx() = tet30[i].xx()*pow(CS,2) + tet30[i].yy()*pow(SIN,2) + 2 * tet30[i].xy()*CS*SIN;   // TauS_rr
 tet3RT1[i].yy() = tet30[i].xx()*pow(SIN,2) + tet30[i].yy()*pow(CS,2) - 2 * tet30[i].xy()*CS*SIN;   // TauS_tt
 tet3RT1[i].xy() = (tet30[i].yy()-tet30[i].xx() )*CS*SIN + tet30[i].xy() * ( pow(CS,2)-pow(SIN,2) ); `

3. Transfering data to cylindrical coordinates and calculation of stress tensor values on the surface (Suppose the tet3RT is the stress tensor in cylindrical coordinates):

`List<symmTensor> tet3RT2 = tet3RT.boundaryField()[patchI];  `

The obtained values are a little different (between 1%-20% difference in different points)! But I don't expect to see even this small difference, because there is no reason for that! I don't know which of them is the exact value!
Do you have any idea?
Note: I know that this problem may be for OpenFOAM and it's not specifically for this solver, but I faced it, when I used this solver, so I decided to ask here. I will appreciate it if you answer me.

**Comment by fppimenta** [comment]
What´s the BC for your tensor?

**Reply by Joe199999** [reply]
Dear Francisco

for both tauS (Cartesian coordinates) and tauSRT (Cylindrical Coordinates) BCs are the same:

for inlet: fixedValue: (0 0 0 0 0 0)
for front and back: empty
for other boundaries (outlet, cylinder surface, up and down boundaries): zerGradient

**Comment by fppimenta** [comment]
Why would you expect to get equal values from the two approaches (in the general case)?

**Reply by Joe199999** [reply]
Because I calculated one thing in both approaches, just the order is different. Could you please explain what the difference is? This is what I want to know. I extracted the stress values on the cylinder surface, then transferred them to the cylindrical coordinates and (2) I transferred the stress tensor to the cylindrical coordinates in the whole domain, then calculated the stress on the cylinder surface.
I think both must get the same results.

**Comment by fppimenta** [comment]
Lets assume your wall boundaries, where the zero-gradient BC is applied. Lets assume a given cell contiguous to the wall. In your approach (2), what you are likely doing is to make the transform of the tensor components at the cell center position and then just copy that cell value to the face (wall), because this is what a zero-gradient BC means. In your approach (1), you transform that same Cartesian tensor components directly at the face location. In general, you should not expect the two approaches to behave the same. 

Note: as far as I remember there are dedicated classes in openfoam that do coordinates transforms. This wont fix your 'problem', but will save you some time doing component-wise transforms.

**Reply by Joe199999** [reply]
Thanks for your kind response.
Actually, I thought that zeroGrad BC calculates filed values on a boundary surface using interpolation!
But I still believe that they should give the same results. Because just the orders are different. What is the difference that I transfer, then put the wall values equal to the neighbor cell, and put the wall values equal to the neighbor cell, then transfer to the cylindrical coordinates? 
So sorry for bothering you by no understanding and asking this question again!
